versioned_docs/version-2021fa/lecture9.md

```diff
@@ -0,0 +1,546 @@
+
+The beginning of your `backend/index.ts` should look like the below. Note the CORS line (line 7) and express static serving line (line 8) we described above and the relevant import statements.
+
+```ts title="backend/index.ts"
+import express from 'express';
+import cors from 'cors';
+import path from 'path';
+import { db, auth } from './firebase-config';
+
+const app = express();
+app.use(cors());
+app.use(express.static(path.join(__dirname, '../../frontend/build')));
+app.use(express.json());
+
+const port = process.env.PORT || 8080;
+
+type Product = {
+  price: string;
+  stocked: boolean;
+  name: string;
+};
+
+type ProductWithID = Product & {
+  id: string;
+};
+
+const productsCollection = db.collection('products');
+
+app.get('/getProducts', async (_, res) => {
+  const products = await productsCollection.get();
+  res.json(
+    products.docs.map((doc): ProductWithID => {
+      const product = doc.data() as Product;
+      return { ...product, id: doc.id };
+    })
+  );
+});
+
+// other routes...
+app.listen(port, () => console.log(`Example app listening on port ${port}!`));
+```
+
+:::note
+Notice we are having `app.listen` listen for either `process.env.PORT` or port 8080. Previously, we hardcoded `8080` for this parameter. `process.env.PORT` will be defined by Heroku and we want the app to listen for requests on that port in the deployed site.
+:::
+
+```ts title="backend/firebase-config.ts"
+import * as admin from 'firebase-admin';
+import { readFileSync } from 'fs';
+import { config } from 'dotenv';
+
+config();
+
+const serviceAccountPath = './firebase-adminsdk.json';
+
+const hydrateServiceAccount = (
+  serviceAccountPath: string
+): admin.ServiceAccount => {
+  const serviceAccount = JSON.parse(
+    readFileSync(serviceAccountPath).toString()
+  );
+  const privateKey = process.env.FIREBASE_PRIVATE_KEY?.replace(/\\n/g, '\n');
+  return { ...serviceAccount, privateKey };
+};
+
+admin.initializeApp({
+  credential: admin.credential.cert(hydrateServiceAccount(serviceAccountPath)),
+});
+
+const db = admin.firestore();
+const auth = admin.auth();
+
+export { db, auth };
+```
+
+#### Frontend
+
+No further setup required in our frontend folder.
+
+#### Root
+
+##### `package.json`
+
+Our root `package.json` will look like the following. We are using yarn workspaces as described [above](./lecture9#root) and the `heroku-postbuild` script will build both workspaces to prep them for push to remote server. (We discussed `yarn workspaces run build` [above](./lecture9#profit)) `heroku-postbuild` is a special command recognized by Heroku to allow you to customize the build process.
+
+```json title="/package.json"
+{
+  "name": "products-app",
+  "version": "1.0.0",
+  "private": true,
+  "scripts": {
+    "heroku-postbuild": "yarn workspaces run build"
+  },
+  "workspaces": ["backend", "frontend"],
+  "license": "MIT",
+  "devDependencies": {
+    "heroku": "^7.59.1"
+  },
+  "resolutions": {
```

Comment:
Let's add a note that heroku cli is bugged at the time of writing, so we have to add this to our root `package.json` and then `yarn install` again as a workaround. Maybe put this part as a FAQ/"it's not working" section so people can try this fix if heroku doesn't work the first time around for them.

Reply:
Normally you don't need this resolutions section at all